r-package/grf/R/causal_tuning.R

```diff
@@ -156,34 +156,97 @@ tune_causal_forest <- function(X, Y, W, Y.hat, W.hat,
     mean(prediction$debiased.error, na.rm = TRUE)
   })
 
+  if (all(is.na(small.forest.errors))) {
+    warning(paste0("Could not tune causal forest because all small forest error estimates were NA.\n",
+                   "Consider increasing argument num.fit.trees."))
+    out <- list("error" = NA, "params" = c(all.params), status = "failure")
+    class(out) <- c("tuning_output")
+    return(out)
+  }
+  if (sd(small.forest.errors) < 1e-10) {
+    warning(paste0("Could not tune causal forest because small forest errors were nearly constant.\n",
+                   "Consider increasing argument num.fit.trees."))
+    out <- list("error" = NA, "params" = c(all.params), status = "failure")
+    class(out) <- c("tuning_output")
+    return(out)
+  }
+
   # Fit the 'dice kriging' model to these error estimates.
   # Note that in the 'km' call, the kriging package prints a large amount of information
   # about the fitting process. Here, capture its console output and discard it.
-  variance.guess <- rep(var(debiased.errors) / 2, nrow(fit.draws))
+  variance.guess <- rep(var(small.forest.errors) / 2, nrow(fit.draws))
   env <- new.env()
   capture.output(env$kriging.model <-
     DiceKriging::km(
       design = data.frame(fit.draws),
-      response = debiased.errors,
+      response = small.forest.errors,
       noise.var = variance.guess
     ))
-  kriging.model <- env$kriging.model
+  model <- env$kriging.model
 
   # To determine the optimal parameter values, predict using the kriging model at a large
-  # number of random values, then select those that produced the lowest error.
+  # number of random values, then select those that produced the lowest predicted error.
   optimize.draws <- matrix(runif(num.optimize.reps * num.params), num.optimize.reps, num.params)
   colnames(optimize.draws) <- names(tuning.params)
-  model.surface <- predict(kriging.model, newdata = data.frame(optimize.draws), type = "SK")
-
+  model.surface <- predict(model, newdata = data.frame(optimize.draws), type = "SK")$mean
   tuned.params <- get_params_from_draw(X, optimize.draws)
-  grid <- cbind(error = model.surface$mean, tuned.params)
-  optimal.draw <- which.min(grid[, "error"])
-  optimal.param <- grid[optimal.draw, ]
 
-  out <- list(
-    error = optimal.param[1], params = c(fixed.params, optimal.param[-1]),
-    grid = grid
+  grid <- cbind(error = c(model.surface), tuned.params)
+  small.forest.optimal.draw <- which.min(grid[, "error"])
+  small.forest.optimal.params <- grid[small.forest.optimal.draw, -1]
+  small.forest.optimal.error <- grid[small.forest.optimal.draw, 1]
```

Comment:
To avoid the possibility selection bias, it'd be better to re-train a small forest at the learned value, and then use that for the comparison with defaults.